docs/docs/guide/project/dependencies.md

```diff
@@ -244,6 +244,55 @@ pod install
+
+**Our recommendation is to trade some build-time convenience for the reliability and determinism of making this decision yourself.** By doing so, Xcode will work more reliably with your project, and your builds will be more predictable.
+
+The simplest approach is to link everything statically in release builds for faster booting and dynamically in debug builds for faster iteration. You can use [dynamic configurations](http://localhost:5173/guide/project/dynamic-configuration.html#reading-the-environment-variables-from-manifests) to manage this at generation-time. You need to include a helper function in your manifest files that returns the appropriate product type based on the environment variable `TUIST_LINKING`:
+
+```swift
+func productType() -> Product {
+    if case let .string(linking) = Environment.linking {
+        return linking == "static" ? .staticFramework : .framework
+    } else {
+        return .framework
+    }
+}
+```
+
+> [!NOTE] PACKAGE PRODUCT TYPE OVERRIDES
+> Tuist allows you to specify the product type for Swift Packages in the `Package.swift` [file](http://localhost:5173/guide/project/dependencies.html#tuist-s-xcodeproj-based-integration). If the integration is done through Xcode's default mechanism, the product type is inferred at build-time.
+
+If making all the targets either static or dynamic is not feasible, you need to be cautious with targets that depend transitively on a static target through dynamic targets. In these cases, Tuist will display a "static side effect" warning to help you identify the issue. These side effects often manifest as increased binary size or, in the worst cases, runtime crashes.
+
+> [!TIP] COMPOSABLE ARCHITECTURE
+> A Swift Package that many projects integrate is [Composable Architecture](https://github.com/pointfreeco/swift-composable-architecture). Due to the the implementation of the packages, linking them statically causes applications to blow up at runtime. To avoid this, you'll have to override the product type for the package in the `Package.swift` file.
+> ```swift
+> #if TUIST
+> import ProjectDescription
+> let packageSettings = PackageSettings(
+>   productTypes: [
+>     "ComposableArchitecture": .framework,
+>     "Dependencies": .framework,
+>     "Clocks": .framework,
+>     "ConcurrencyExtras": .framework,
+>     "CombineSchedulers": .framework,
+>     "IdentifiedCollections": .framework,
+>     "OrderedCollections": .framework,
+>     "_CollectionsUtilities": .framework,
+>     "DependenciesMacros": .framework,
+>     "SwiftUINavigationCore": .framework,
+>     "Perception": .framework,
+>     "CasePaths": .framework,
+>     "CustomDump": .framework,
+>     "XCTestDynamicOverlay": .framework
+>. ]
+>)
+>#endif
+>```
```

Comment:
What about adding this custom override directly in Tuist?

Reply:
That's an anti-pattern that we should revert at some point, so I'd rather not do it. Tuist should be package-agnostic.

Reply:
I think TCA is important enough that I would consider this. If we know the library will blow up in runtime, then we shouldn't let the project be generated without any warnings.

Reply:
The project that we produce is valid. The configuration is wrong, and we've learned from Xcode what happens when you magically resolve configurations for users. Let's not go down that path.

Reply:
I disagree. Tuist does what it can with the information SPM provides it with. If SPM does a bad design decision, like it's the case here, it should not be our responsibility to make assumptions and decisions on behalf of users like they do. Because when we or someone does, things like [this](https://github.com/tuist/tuist/pull/6244), [this](https://github.com/CocoaPods/CocoaPods/issues/4101), or [this](https://github.com/bazelbuild/rules_apple/issues/1938).

So while it sounds amazing that Tuist would magically make things work, this has a [proven cost for users](https://docs.tuist.io/guide/introduction/cost-of-convenience.html), and a cost for us, that all of a sudden need to monitor what those hardcoded packages are doing, and make sure that we iterate on the auto-applied configurations as those packages evolve, or potentially having to make exceptions because [maybe the configuration that we applied](https://github.com/tuist/tuist/issues/6243) is not the one that users spend.

I'd rather invest the time in providing the users with the right resources and documentation to make this decision, to have a very explicit and static graph that makes things compile fast and Xcode work reliably. Sure, users would need to read a bit, but I'd rather adopt the language here of "translating a complicated subject" than venturing through the same path that Xcode and SPM took and that motivated the creation of Tuist.

Reply:
-ObjC and static vs dynamic are slightly different issues. 

The necessity to add ObjC in case of Composable Architecture should be considered a bug. SPM doesn't add it and nor should our users. Since this is a bug (or peculiarity of our integration if you will), we might want to consider adding it, but ideally, we figure out why it's needed in the first place. 

The static vs dynamic issue is currently our design decision. While I agree with you on the cost of convenience, I don't think it's beneficial to go against the whole SPM ecosystem with our approach. Imho, we should respect the "automatic" SPM option because that's what users and package authors expect.

We could make the default dynamic as libraries work usually better when linked that way, but then we'd be even further from the whole ecosystem as SPM's default is static.

In the end, we want Tuist users to use our integration because it allows us to optimise those projects. We needed to diverge from SPM here and there to allow those optimisations, but not supporting automatic type is an unncessary divergence that causes unnecessary friction from Tuist users to adopt our approach. 

I'd have questions about using Tuist's approach if it means I need to suddenly configure linking for tens of package targets because Tuist diverges from the SPM ecosystem.

FWIW, aligning Tuist with SPM as much as possible is also beneficial for Tuist the company since a big chunk of the revenue comes from caching. If we make XcodeProj-based integration more difficult than necessary, we will also put away companies who fail at migrating to our approach due to the discrepancies.

Reply:
> -ObjC and static vs dynamic are slightly different issues.

They are not. Static linking overoptimizes and forces the build setting `—ObjC`. Dynamic linking does not, and therefore, `—ObjC`  is not needed. SPM can't know whether `—ObjC` is needed, nor can Bazel and CocoaPods, so they decide for dynamic on behalf of users, and things *just work*™️.

I don't think we should replicate the SPM's "automatic" linking because:

1. **Costly:** It's an endless game of mimicking SPM's logic. The cost is relatively high already, so we'd make it even higher, and waste time we'd otherwise spend building valuable features.
2. **Assumptions:** It's a game that requires looking at the sources and the graph, potentially making wrong assumptions and slowing down project generation.
3. **Cargo-culting of bad design:** It's a design principle that led Xcode to where it is today and motivated the creation of Tuist. We must prevent Tuist from ending in the same position and letting bad decisions leak through our design.

Users on issues have proven that given the right guidelines, they not only fix their issues but get to think about their graph, which has deep implications for the UX of their apps (e.g., boot time). SPM deciding that all the Composable targets are dynamic because it's the way to make things work might result in apps taking much longer to boot.  If Tuist helps organizations build better apps, this goes in a completely different direction. Proper guidelines never existed.

Reply:
> SPM can't know whether —ObjC is needed, nor can Bazel and CocoaPods, so they decide for dynamic on behalf of users, and things just work™️.

SPM does _not_ add `-ObjC` flag at all afaik. Which is why it should be considered a bug or at least an unwanted discrepancy. If that's only because that flag is not necessary if a library is dynamic, then that's a different story, though.

> Users on issues have proven that given the right guidelines, they not only fix their issues but get to think about their graph

But there might be lots of users that don't figure it out and just keep using the native SPM integration. Of course, that might be just my feeling, but I assume not every developer ends up filing an issue or reaches out in Slack.

> Assumptions: It's a game that requires looking at the sources and the graph, potentially making wrong assumptions and slowing down project generation.

We should be able to look at a graph only and make a target dynamic if users would hit the duplicate products warning. We already have a logic in a form of a linter, so what we'd need to do is in those scenarios mutate the graph. We could add guidelines for how we resolve the product type the same way we're doing it for the very arbitrary default of a static framework. We'd still let developers make the linking explicit – which is what SPM doesn't allow you as of now.

> Costly: It's an endless game of mimicking SPM's logic. The cost is relatively high already, so we'd make it even higher, and waste time we'd otherwise spend building valuable features.

Costly is also our support that we have to give on the discrepancies – but maybe it will get better with better guidelines. 

> Users on issues have proven that given the right guidelines, they not only fix their issues but get to think about their graph, which has deep implications for the UX of their apps (e.g., boot time).

You mentioned on Slack that you would consider making `.dynamic` the default linking type because libraries work better with it. That will by default worsen the boot time of every app – and most users keep tool's defaults unless they're forced to change it. If we're forced to go this route, I'd definitely lean to respecting the `.automatic` SPM option.